Generator: Add full auroLabs docs (#95)
AlaskaAirlines/WC-Generator: docs/customElementCss.md

# CSS styles and custom elements

CSS with custom elements opens a new bucket of opportunity. A few benefits that Auro takes direct advantage of are encapsulation, `:host`, and constructable stylesheets. Let's dive into these technologies and understand how best to use them.

## Encapsulation

> One of the key features of the Web Components standard is the ability to create custom elements that encapsulate your functionality on an HTML page, rather than having to make do with a long, nested batch of elements that together provide a custom page feature.

-- [Using custom elements - MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements)

> So what does that mean? A definition that makes sense to me is, _" the succinct expression or depiction of the essential features of something."_ In essence, that is also the definition of a custom element. It is the encapsulation with the [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM) that allows web developers the unique opportunity to create a component that will not pollute anything outside the scope of the element and has very few things that have any influence on it. The key ones being global font and color styles, and [CSS custom properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties).

Reviewer: ```suggestion
So what does that mean? One definition is, _" the succinct expression or depiction of the essential features of something."_ In essence, that is also the definition of a custom element. It is the encapsulation with the [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM) that allows web developers the unique opportunity to create a component that will not pollute anything outside the scope of the element and is mostly isolated from the outside. The key properties that affect a custom element from the outside (or "pierce the shadow DOM") are global font and color styles, and [CSS custom properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties).
```

> The true benefit with encapsulation of CSS and custom elements is that you do not need to be concerned about global styles having side effects on your element when used in any context of any website. While naming conventions and using parent selectors have become a staple of web development over the years, these techniques are not necessary for UI stability in custom elements. But they do come in handy for maintaining the sanity of your CSS. To that end, Auro has a series of [CSS development conventions](https://auro.alaskaair.com/webcorestylesheets/conventions) that we highly recommend.

Reviewer: ```suggestion
The true benefit with encapsulation of CSS and custom elements is that you do not need to be concerned about global styles having side effects on your element when used in any context of any website. While naming conventions and using parent selectors have become a staple of web development over the years, these techniques are not necessary for UI stability in custom elements. However, they do come in handy for maintaining the sanity of your CSS. To that end, Auro has a series of [CSS development conventions](https://auro.alaskaair.com/webcorestylesheets/conventions) that we highly recommend.
```

In practical terms, what does that mean? In other front-end frameworks, it is commonplace to wrap the scope of the component's HTML in a `div` and add a class. You could look at that outer `div` as the _host_ of the component.

```js
// React w/JSX
render() {
  return (
    <div className="my-element">
      ...
    </div>
  )
}
```

With custom elements, there is no need to wrap all the HTML inside an outer element to create that host wrapper. The custom element itself is that wrapper and any styles using the `:host` selector will be applied to the whole element. The following example is a HTML template from LitElement using a [tagged template literal](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates). This is NOT JSX.

```js
// LitElement w/tagged template literal
render() {
  return html`
    <div>
      ...
    </div>
  `;
}
```

### Parental awareness

Aside from not needing to wrap all the HTML in yet another `div`, another thing that we get that is a super-power IMHO is parental awareness. What does that mean? Well, a custom element's CSS is fully aware of any changes to the parent custom element. E.g. `<my-element>`.

> Where this comes in VERY handy is with applying stylistic changes to your custom element without having to create properties and manage the application and removal of CSS within the scope of the element. The idea is to allow a user to add an attribute to the custom element tag and the element will just respond.

Reviewer: It would be good to note here that the property needs to be reflected to attributes if you are using them as styling hooks to support frameworks that only set properties (e.g. Svelte & Preact).

Author: I know what you are referencing, but I don't remember the fix?

Author: I added a comment to the page to come back and address where you made this mention.

```css
.content ::slotted(p) {
  margin-top: var(--auro-size-none);
}

.content ::slotted(p:last-child) {
  margin-bottom: var(--auro-size-none);
}
```

The styles inside the scope of the custom element do not influence the `<p>` element that may be inheriting styles from a global stylesheet. To counter this, using the `::slotted` selector the author of the element can exert control over the slotted `<p>` element.

This is pretty powerful, but there are limitations. The `::slotted` selector [does not influence nested elements](https://stackoverflow.com/questions/61626493/slotted-css-selector-for-nested-children-in-shadowdom-slot).

## Shadow parts

This is an exciting new development with CSS and the growing support for custom element customization.

> The ::part CSS pseudo-element represents any element within a shadow tree that has a matching part attribute.

-- [::part](https://developer.mozilla.org/en-US/docs/Web/CSS/::part)

What do shadow parts give us? Simply put, this allows the author of the element to define specific parts of the HTML template that will allow users to target those parts and apply CSS that will affect the UI inside the shadow DOM.

> Hoping you have read through everything up to this point, you may see where this is yet another super-power. What we know so far is that CSS in the shadow DOM stays in the shadow DOM. We also know that CSS outside the scope of the shadow DOM has little effect on any custom element. While influence on slotted content or using the `::slotted` selector gives users some opportunity, `::part` is the most powerful of all.

Reviewer: ```suggestion
If you have read through everything up to this point, you may see where this is yet another super-power. What we know so far is that CSS in the shadow DOM stays in the shadow DOM. We also know that CSS outside the scope of the shadow DOM has little effect on any custom element. While influence on slotted content or using the `::slotted` selector gives users some opportunity, `::part` is the most powerful of all.
```
